File: evaluation/S0/javascript/Hard/2767-count-the-number-of-winning-sequences.js
// Source: https://leetcode.com/problems/count-the-number-of-winning-sequences/   |   Difficulty: Hard
//
// Problem Description:
// Alice and Bob are playing a fantasy battle game consisting of n rounds where they summon one of three magical creatures each round: a Fire Dragon, a Water Serpent, or an Earth Golem. In each round, players simultaneously summon their creature and are awarded points as follows:
//
//
// 	If one player summons a Fire Dragon and the other summons an Earth Golem, the player who summoned the Fire Dragon is awarded a point.
// 	If one player summons a Water Serpent and the other summons a Fire Dragon, the player who summoned the Water Serpent is awarded a point.
// 	If one player summons an Earth Golem and the other summons a Water Serpent, the player who summoned the Earth Golem is awarded a point.
// 	If both players summon the same creature, no player is awarded a point.
//
//
// You are given a string s consisting of n characters 'F', 'W', and 'E', representing the sequence of creatures Alice will summon in each round:
//
//
// 	If s[i] == 'F', Alice summons a Fire Dragon.
// 	If s[i] == 'W', Alice summons a Water Serpent.
// 	If s[i] == 'E', Alice summons an Earth Golem.
//
//
// Bob’s sequence of moves is unknown, but it is guaranteed that Bob will never summon the same creature in two consecutive rounds. Bob beats Alice if the total number of points awarded to Bob after n rounds is strictly greater than the points awarded to Alice.
//
// Return the number of distinct sequences Bob can use to beat Alice.
//
// Since the answer may be very large, return it modulo 109 + 7.
//
// Example:
// Input: s = "FFF"
//
// Output: 3
//
// Explanation:
//
// Bob can beat Alice by making one of the following sequences of moves: "WFW", "FWF", or "WEW". Note that other winning sequences like "WWE" or "EWW" are invalid since Bob cannot make the same move twice in a row.
//
// Constraints:
// 1 <= s.length <= 1000
// 	s[i] is one of 'F', 'W', or 'E'.
//

/**
 * @param {string} s
 * @return {number}
 */
var countWinningSequences = function(s) {
    const MOD = 1e9 + 7;
    const n = s.length;
    const dp = Array.from({ length: n + 1 }, () => Array(3).fill(0).map(() => Array(2 * n + 1).fill(0)));
    const charToIndex = { 'F': 0, 'W': 1, 'E': 2 };
    dp[0][0][n] = 1;
    dp[0][1][n] = 1;
    dp[0][2][n] = 1;
    for (let i = 1; i <= n; i++) {
        const aliceMove = charToIndex[s[i - 1]];
        for (let bobMove = 0; bobMove < 3; bobMove++) {
            for (let scoreDiff = 0; scoreDiff <= 2 * n; scoreDiff++) {
                for (let prevBobMove = 0; prevBobMove < 3; prevBobMove++) {
                    if (bobMove === prevBobMove) continue;
                    let newScoreDiff = scoreDiff;
                    if ((bobMove === 0 && aliceMove === 2) || (bobMove === 1 && aliceMove === 0) || (bobMove === 2 && aliceMove === 1)) {
                        newScoreDiff++;
                    } else if ((aliceMove === 0 && bobMove === 2) || (aliceMove === 1 && bobMove === 0) || (aliceMove === 2 && bobMove === 1)) {
                        newScoreDiff--;
                    }
                    if (newScoreDiff >= 0 && newScoreDiff <= 2 * n) {
                        dp[i][bobMove][newScoreDiff]
                            = (dp[i][bobMove][newScoreDiff] + dp[i - 1][prevBobMove][scoreDiff]) % MOD;
                    }
                }
            }
        }
    }       
    let result = 0;
    for (let bobMove = 0; bobMove < 3; bobMove++) {
        for (let scoreDiff = n + 1; scoreDiff <= 2 * n; scoreDiff++) {
            result = (result + dp[n][bobMove][scoreDiff]) % MOD;
        }
    }
    return result;
};

File: evaluation/S2/c/Hard/2767-count-the-number-of-winning-sequences.c
// Source: https://leetcode.com/problems/count-the-number-of-winning-sequences/   |   Difficulty: Hard
//
// Problem Description:
// Alice and Bob are playing a fantasy battle game consisting of n rounds where they summon one of three magical creatures each round: a Fire Dragon, a Water Serpent, or an Earth Golem. In each round, players simultaneously summon their creature and are awarded points as follows:
//
//
// 	If one player summons a Fire Dragon and the other summons an Earth Golem, the player who summoned the Fire Dragon is awarded a point.
// 	If one player summons a Water Serpent and the other summons a Fire Dragon, the player who summoned the Water Serpent is awarded a point.
// 	If one player summons an Earth Golem and the other summons a Water Serpent, the player who summoned the Earth Golem is awarded a point.
// 	If both players summon the same creature, no player is awarded a point.
//
//
// You are given a string s consisting of n characters 'F', 'W', and 'E', representing the sequence of creatures Alice will summon in each round:
//
//
// 	If s[i] == 'F', Alice summons a Fire Dragon.
// 	If s[i] == 'W', Alice summons a Water Serpent.
// 	If s[i] == 'E', Alice summons an Earth Golem.
//
//
// Bob’s sequence of moves is unknown, but it is guaranteed that Bob will never summon the same creature in two consecutive rounds. Bob beats Alice if the total number of points awarded to Bob after n rounds is strictly greater than the points awarded to Alice.
//
// Return the number of distinct sequences Bob can use to beat Alice.
//
// Since the answer may be very large, return it modulo 109 + 7.
//
// Example:
// Input: s = "FFF"
//
// Output: 3
//
// Explanation:
//
// Bob can beat Alice by making one of the following sequences of moves: "WFW", "FWF", or "WEW". Note that other winning sequences like "WWE" or "EWW" are invalid since Bob cannot make the same move twice in a row.
//
// Constraints:
// 1 <= s.length <= 1000
// 	s[i] is one of 'F', 'W', or 'E'.
//
// Helpful references (internal KB):
// - The Inclusion-Exclusion Principle (string, counting)
//   • When to use: Use when you need to determine the frequency of characters or substrings within a string. It is essential for tasks like anagram detection or checking character distribution.
//   • Idea: This technique involves iterating through a string to tally the occurrences of its constituent elements, typically characters or words. It generally operates in O(N) time complexity, where N is the string's length.
//   • Invariants: The count for each character reflects its total occurrences processed so far.; All characters encountered up to the current position have been tallied.
//   • Tips: Use a hash map (dictionary) for arbitrary character sets.; Use an array for fixed-size character sets (e.g., ASCII, lowercase English).
//   • Pitfalls: Forgetting to handle case sensitivity or ignore non-alphanumeric characters.; Incorrectly handling Unicode characters with fixed-size arrays.
// - palindrome-partitioning (string, two-pointers, simulation)
//   • When to use: When a string needs to be decomposed into all possible sets of palindromic substrings. It's suitable for problems requiring exhaustive enumeration of such partitions.
//   • Idea: This algorithm finds all possible ways to partition a given string such that every substring in the partition is a palindrome. It typically uses a recursive exploration approach combined with a two-pointer check for palindrome validation, resulting in an exponential time complexity.
//   • Invariants: At any point, the `step` list contains a valid palindromic partition of the prefix processed so far.; The `s` parameter in the recursive function always represents the unpartitioned suffix of the original string.
//   • Tips: Implement the `isPalindrome` check efficiently using two pointers for O(length) time.; Use a recursive helper function to explore all possible partition points.
//   • Pitfalls: Incorrectly handling substring boundaries, leading to off-by-one errors or incorrect partitions.; Forgetting to backtrack (remove the last added substring) after a recursive call returns.
// - Z-function and its calculation (string, array, two-pointers)
//   • When to use: Use when you need to find the length of the longest common prefix between a string and all its suffixes efficiently. It is fundamental for various string matching and pattern searching problems.
//   • Idea: The Z-function computes an array where each element z[i] is the length of the longest common prefix between the original string and its suffix starting at index i. This algorithm calculates the entire Z-array in O(N) time complexity.
//   • Invariants: For all k < i, z[k] has been correctly computed.; The interval [l, r) always represents the rightmost segment match found so far.
//   • Tips: Leverage previously computed Z-values to avoid redundant comparisons.; Maintain a [l, r) window representing the rightmost segment match found so far.
//   • Pitfalls: Incorrectly handling the base case z[0] (usually set to 0).; Off-by-one errors when calculating the length of the prefix or updating the window.
//
// Notes:
// - Keep the given function signature unchanged.
// - Use the C standard library only; no external packages.
// - Do not change input/output formats or return value semantics.
// - Conformance: ISO C (C89/C99/C11). Do NOT use GNU C extensions (nested functions, statement expressions, typeof, __auto_type, etc.). Avoid VLAs and non-standard builtins.
// - If you introduce any helper functions, include their full implementations in this file; no hidden helpers.

int countWinningSequences(char* s) {
    int mod = 1e9 + 7;
    int n = strlen(s);
    long long dp[1001][3][3]; // dp[i][j][k] = number of ways for Bob to win up to round i, with Bob's last move j and Alice's last move k

    // Initialize dp array
    memset(dp, 0, sizeof(dp));
    
    // Base case: no rounds, no points
    dp[0][0][0] = 1; // Bob can start with any move

    for (int i = 1; i <= n; i++) {
        char aliceMove = s[i - 1];
        for (int bobMove = 0; bobMove < 3; bobMove++) {
            for (int aliceLastMove = 0; aliceLastMove < 3; aliceLastMove++) {
                if (bobMove == aliceLastMove) continue; // Bob cannot repeat the same move
                int pointsBob = 0, pointsAlice = 0;
                if ((bobMove == 0 && aliceMove == 'E') || (bobMove == 1 && aliceMove == 'F') || (bobMove == 2 && aliceMove == 'W')) {
                    pointsBob++;
                } else if ((aliceLastMove == 0 && aliceMove == 'E') || (aliceLastMove == 1 && aliceMove == 'F') || (aliceLastMove == 2 && aliceMove == 'W')) {
                    pointsAlice++;
                }
                dp[i][bobMove][aliceLastMove] = (dp[i][bobMove][aliceLastMove] + dp[i - 1][bobMove][aliceLastMove]) % mod;
            }
        }
    }

    long long result = 0;
    for (int bobMove = 0; bobMove < 3; bobMove++) {
        for (int aliceLastMove = 0; aliceLastMove < 3; aliceLastMove++) {
            if (dp[n][bobMove][aliceLastMove] > dp[n][aliceLastMove][bobMove]) {
                result = (result + dp[n][bobMove][aliceLastMove]) % mod;
            }
        }
    }

    return result;      
}
